### PR TITLE
Update `WindowInfo.containerSize` on window resize to keep it correct when window is moved between displays

### LINKED ISSUE #4697: Moving the app to another monitor breaks dropdowns
**Describe the bug**
After moving the app to another monitor, dropdown menu positioning allows overflow outside the window

**Affected platforms**
- Desktop (macOS)

**Versions**
- Kotlin version*: 1.9.23
- Compose Multiplatform version*: 1.6.2
- OS version(s)* (required for Desktop and iOS issues): MacOS 14
- OS architecture (x86 or arm64): Arm64
- Device (model or simulator for iOS issues): Mac Pro M1
- JDK (for desktop issues): Temurin 17

**To Reproduce**
Steps and/or the code snippet to reproduce the behavior:
1. Download https://github.com/Syer10/compose-window-issue
2. Open the app
3. Move the app to another monitor
4. Click the button to open the dropdown
5. See wrong position

**Expected behavior**
Dropdown is moved to show the whole dropdown instead of only the first part

**Screenshots**
image

**Additional context**
Only happens on MacOS

### PR DESCRIPTION
**Pull request by m-sasha**
`ComposeContainter` currently registers the listener for window size and position on the `windowContainer`, which is a `JLayeredPane`. It ends up not being called at all when the window is resized or moved. According to @MatkovIvan, however, this is intentional and merely poorly named (e.g. `onChangeWindowPosition` is meant to be called when the `windowContainer`'s position changes, not the window's).

Because of this, when a window is moved to another screen with a different density, the size we store for it in `WindowInfo.containerSize` isn't updated. This, in turn, causes issues with popup positioning.

This PR registers an additional size listener on the window, in order to update `WindowInfo.containerSize` when the window size changes. It also renames all the related listener functions.

Fixes https://github.com/JetBrains/compose-multiplatform/issues/4697 
Possibly fixes additional issues where the window size was not being updated after moving it to a different screen.

## Testing
Tested manually with this code:
```
@Composable
fun TestButton(text: String) {
    val displayHover = remember { mutableStateOf(false) }

    Column(
        modifier = Modifier
    ) {
        Button(
            onClick = {
                displayHover.value = true
            },
            modifier = Modifier.pointerHoverIcon(PointerIcon.Hand)
        ) {
            Text(text)
        }
        DropdownMenu(
            displayHover.value,
            onDismissRequest = {
                displayHover.value = false
            },
            modifier = Modifier.background(MaterialTheme.colors.background),
        ) {
            Row(Modifier.width(200.dp), horizontalArrangement = Arrangement.SpaceBetween) {
                Text(
                    text = "My tooltip",
                )
                Text(
                    text = "2 tooltip",
                )
            }

        }
    }
}

fun main() = singleWindowApplication {
    MaterialTheme {
        Box(Modifier.fillMaxSize(), contentAlignment = Alignment.TopEnd) {
            TestButton("Hello, World!")
        }
    }
}
```

To simulate a 2nd screen, I've used this app: https://github.com/Stengo/DeskPad and configured the 2nd screen to be of a very low resolution (such that the density is 1.0).

This could be tested by QA

### Fixes - Desktop

- Fix dropdown/popup positioning when a window is moved to a screen with a different density

### CODE REVIEW

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt [hunk 1]
```diff
@@ -111,9 +111,6 @@ internal class ComposeContainer(
                 error("Customizing windowContainer cannot be used with LayerType.OnSameCanvas")
             }
 
-            _windowContainer?.removeComponentListener(this)
-            value.addComponentListener(this)
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:115]: It's incorrect. `windowContainer` is a reference for size, not a window. We can register additional event, but not replace existing one.

Use case: From #992 but setup `windowContainer` as only part of the window. We need to improve out test cases if it's not covered yet to prevent breaking this.

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt (hunk 1 above)
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:115]: See latest commit.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Popup.skiko.kt [hunk 2]
```diff
@@ -512,7 +512,10 @@ private fun rememberPopupMeasurePolicy(
                 -platformInsets.top.roundToPx()
             )
             val positionInWindow = popupPositionProvider.calculatePosition(
-                boundsWithoutInsets, sizeWithoutInsets, layoutDirection, contentSize
```
Comment (MatkovIvan) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Popup.skiko.kt:515]: Please revert unrelated change

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt [hunk 3]
```diff
@@ -139,6 +139,36 @@ internal class ComposeContainer(
...
+    }
+
+    /**
+     * The listener to [_windowContainer] size and position changes.
+     */
+    private val windowContainerComponentListener = object : ComponentAdapter() {
+        override fun componentResized(e: ComponentEvent?)  {
+            if (!container.isDisplayable) return
+
+            mediator.onChangeComponentSize()
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: It's still incorrent. `onChangeWindowSize` **must** be called here - we need to update the size here. Look, `windowContainer`'s size is used in `onChangeWindowSize`

**Review by MatkovIvan — CHANGES_REQUESTED** [review]

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Popup.skiko.kt (hunk 2 above)
Comment (m-sasha) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/window/Popup.skiko.kt:515]: It's not entirely unrelated, and it improves readability, and I don't want to waste time creating a separate PR for such a tiny change.

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt (hunk 3 above)
Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: Right, I'll move 
```
        windowContext.setContainerSize(windowContainer.sizeInPx)
```
here instead.

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: Well, actually not move, but copy.

Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: this line not enough, layers must be updated too

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: So the names of the functions on the layers and their javadoc are wrong?

Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: Window size reference for Compose is windowContainer, not real window. Compose shouldn't depend on "real" window for sizing (it does depend for things like window state/focus/etc)
We can rename window -> windowContainer in a few places to improve readability.

For the specific case, I'd avoid _removing_ re-checking size from callbacks and limit this PR only to _adding_ new callbacks

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: I don't think that's correct for all cases. For example you would want to center dialogs in the window rather than the window container.

Comment (MatkovIvan) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: That's exactly the case! window container IS the referece for that

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt [hunk 4]
```diff
@@ -139,6 +139,28 @@ internal class ComposeContainer(
...
+     * The listener to [window] size changes.
+     */
+    private val windowSizeListener = object : ComponentAdapter() {
+        override fun componentResized(e: ComponentEvent?) {
+            if (!container.isDisplayable) return
+
+            // When the window is moved to a display with a different density, componentResized is
+            // called on the window, but not on `windowContainer`, so we need to update the size
+            // here too.
+            windowContext.setContainerSize(windowContainer.sizeInPx)
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt]: Please call `onWindowContainerSizeChanged` instead - layers should be updated too (for dialog centering for example).
Please recreck your fix with all layer types (canvas, component, window)

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt [hunk 5]
```diff
@@ -174,25 +196,13 @@ internal class ComposeContainer(
...
         layers.fastForEach(DesktopComposeSceneLayer::close)
     }
 
-    override fun componentResized(e: ComponentEvent?)  {
-        onChangeWindowSize()
-
-        // Sometimes Swing displays interop views in incorrect order after resizing,
-        // so we need to force re-validate it.
-        container.validate()
-        container.repaint()
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:188]: Why do you remove this workaround? It's somehow fixed in different place?

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:188]: I put it back.

**Review by MatkovIvan — APPROVED** [review]

File: compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt [hunk 6]
```diff
@@ -139,6 +139,24 @@ internal class ComposeContainer(
...
+        override fun componentMoved(e: ComponentEvent?) = onWindowContainerPositionChanged()
+    }
+
+    /**
+     * The listener to [window] size changes.
+     */
+    private val windowSizeListener = object : ComponentAdapter() {
+        // When the window is moved to a display with a different density, componentResized is
+        // called on the window, but not on `windowContainer`, so we need to update the size
+        // here too.
```
Comment (MatkovIvan) [inline compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:156]: Is it possible to have a test for this case?

Comment (m-sasha) [reply compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/scene/ComposeContainer.desktop.kt:156]: I can't think of an easy way.
